File: backend/src/routes/requests.ts
import { Router } from 'express';
import { pool } from '../db/pool';

const router = Router();

type RequestItemInput = {
  partNumber?: string | number;
  name?: string;
  quantity?: number;
  unit?: string;
  brand?: string;
  model?: string;
  serialNumber?: number;
  scheme?: string;
  posScheme?: string;
  material?: string;
  comment?: string;
  unitPrice?: number;
  totalPrice?: number;
  productId?: number;
};

type RequestInput = {
  idRequest?: number;
  typeRequest?: string;
  datetimeComing?: string;
  datetimeDelivery?: string;
  status?: string;
  totalPrice?: number;
  items?: RequestItemInput[];
};

router.get('/', async (_req, res, next) => {
  try {
    const { rows: requestRows } = await pool.query(
      `select
         r.id,
         r.id_request as "idRequest",
         r.type_request as "typeRequest",
         rt.description as "typeDescription",
         r.datetime_coming as "datetimeComing",
         r.datetime_delivery as "datetimeDelivery",
         r.status,
         rs.description as "statusDescription",
         r.total_price as "totalPrice"
       from requests r
       left join request_types rt on rt.code = r.type_request
       left join request_statuses rs on rs.code = r.status
       order by r.datetime_coming desc`
    );

    const { rows: itemRows } = await pool.query(
      `select
         id,
         part_number as "partNumber",
         name,
         quantity,
         unit,
         brand,
         model,
         serial_number as "serialNumber",
         scheme,
         pos_scheme as "posScheme",
         material,
         comment,
         unit_price as "unitPrice",
         total_price as "totalPrice",
         request_id as "requestId",
         product_id as "productId"
       from request_items`
    );

    const itemsByRequest = itemRows.reduce<Record<number, typeof itemRows>>(
      (acc, item) => {
        const key = item.requestId;
        if (!key) {
          return acc;
        }
        acc[key] = acc[key] ? [...acc[key], item] : [item];
        return acc;
      },
      {}
    );

    const payload = requestRows.map((request) => ({
      ...request,
      items: itemsByRequest[request.id] ?? []
    }));

    res.json(payload);
  } catch (error) {
    next(error);
  }
});

router.post('/', async (req, res, next) => {
  const {
    idRequest,
    typeRequest,
    datetimeComing,
    datetimeDelivery,
    status,
    totalPrice,
    items
  }: RequestInput = req.body ?? {};

  if (!idRequest || Number.isNaN(Number(idRequest))) {
    res.status(400).json({ message: 'Field "idRequest" must be a number' });
    return;
  }

  if (!datetimeComing) {
    res
      .status(400)
      .json({ message: 'Field "datetimeComing" must be a valid ISO date string' });
    return;
  }

  const parsedComing = new Date(datetimeComing);
  if (Number.isNaN(parsedComing.valueOf())) {
    res
      .status(400)
      .json({ message: 'Field "datetimeComing" must be a valid ISO date string' });
    return;
  }

  let parsedDelivery: Date | null = null;
  if (datetimeDelivery) {
    const date = new Date(datetimeDelivery);
    if (Number.isNaN(date.valueOf())) {
      res.status(400).json({
        message: 'Field "datetimeDelivery" must be a valid ISO date string'
      });
      return;
    }
    parsedDelivery = date;
  }

  const client = await pool.connect();

  try {
    await client.query('begin');

    const { rows: requestRows } = await client.query(
      `insert into requests (
        id_request,
        type_request,
        datetime_coming,
        datetime_delivery,
        status,
        total_price
      ) values ($1,$2,$3,$4,$5,$6)
      returning
        id,
        id_request as "idRequest",
        type_request as "typeRequest",
        datetime_coming as "datetimeComing",
        datetime_delivery as "datetimeDelivery",
        status,
        total_price as "totalPrice"`,
      [
        idRequest,
        typeRequest ?? null,
        parsedComing.toISOString(),
        parsedDelivery ? parsedDelivery.toISOString() : null,
        status ?? null,
        totalPrice ?? null
      ]
    );

    const createdRequest = requestRows[0];
    const createdItems: unknown[] = [];

    if (Array.isArray(items) && items.length > 0) {
      for (const item of items) {
        if (!item?.name) {
          throw new Error('Each request item must include "name"');
        }

        const itemPartNumber =
          item.partNumber === undefined || item.partNumber === null
            ? null
            : typeof item.partNumber === 'number'
              ? item.partNumber.toString()
              : typeof item.partNumber === 'string'
                ? item.partNumber.trim() || null
                : null;

        const result = await client.query(
          `insert into request_items (
            part_number,
            name,
            quantity,
            unit,
            brand,
            model,
            serial_number,
            scheme,
            pos_scheme,
            material,
            comment,
            unit_price,
            total_price,
            request_id,
            product_id
          ) values ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15)
          returning
            id,
            part_number as "partNumber",
            name,
            quantity,
            unit,
            brand,
            model,
            serial_number as "serialNumber",
            scheme,
            pos_scheme as "posScheme",
            material,
            comment,
            unit_price as "unitPrice",
            total_price as "totalPrice",
            request_id as "requestId",
            product_id as "productId"`,
          [
            itemPartNumber,
            item.name,
            item.quantity ?? null,
            item.unit ?? null,
            item.brand ?? null,
            item.model ?? null,
            item.serialNumber ?? null,
            item.scheme ?? null,
            item.posScheme ?? null,
            item.material ?? null,
            item.comment ?? null,
            item.unitPrice ?? null,
            item.totalPrice ?? null,
            createdRequest.id,
            item.productId ?? null
          ]
        );

        createdItems.push(result.rows[0]);
      }
    }

    await client.query('commit');

    res.status(201).json({
      ...createdRequest,
      items: createdItems
    });
  } catch (error) {
    await client.query('rollback');
    next(error);
  } finally {
    client.release();
  }
});

export default router;

File: backend_flask/app/routes/requests.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List

from flask import jsonify, request
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import joinedload

from ..database import db
from ..models import Request, RequestItem
from . import api_bp


def serialize_request_item(item: RequestItem) -> Dict[str, Any]:
    return {
        "id": item.id,
        "partNumber": item.part_number,
        "name": item.name,
        "quantity": item.quantity,
        "unit": item.unit,
        "brand": item.brand,
        "model": item.model,
        "serialNumber": item.serial_number,
        "scheme": item.scheme,
        "posScheme": item.pos_scheme,
        "material": item.material,
        "comment": item.comment,
        "unitPrice": item.unit_price,
        "totalPrice": item.total_price,
        "requestId": item.request_id,
        "productId": item.product_id,
    }


def serialize_request(req: Request) -> Dict[str, Any]:
    return {
        "id": req.id,
        "idRequest": req.id_request,
        "typeRequest": req.type_request,
        "typeDescription": req.type.description if req.type else None,
        "datetimeComing": req.datetime_coming.isoformat(),
        "datetimeDelivery": req.datetime_delivery.isoformat() if req.datetime_delivery else None,
        "status": req.status,
        "statusDescription": req.status_rel.description if req.status_rel else None,
        "totalPrice": req.total_price,
        "items": [serialize_request_item(item) for item in req.items],
    }


@api_bp.get("/requests")
def list_requests():
    requests = (
        Request.query.order_by(Request.datetime_coming.desc())
        .options(joinedload(Request.items), joinedload(Request.type), joinedload(Request.status_rel))
        .all()
    )
    return jsonify([serialize_request(req) for req in requests])


def parse_iso_datetime(value: str, field_name: str) -> datetime:
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:  # pragma: no cover - simple parsing guard
        raise ValueError(f'Field "{field_name}" must be a valid ISO string') from exc


@api_bp.post("/requests")
def create_request():
    payload = request.get_json(silent=True) or {}

    id_request = payload.get("idRequest")
    datetime_coming = payload.get("datetimeComing")

    if id_request is None or not isinstance(id_request, int):
        return jsonify({"message": 'Field "idRequest" must be an integer'}), 400

    if not datetime_coming or not isinstance(datetime_coming, str):
        return jsonify({"message": 'Field "datetimeComing" must be an ISO string'}), 400

    try:
        parsed_coming = parse_iso_datetime(datetime_coming, "datetimeComing")
        parsed_delivery = (
            parse_iso_datetime(payload["datetimeDelivery"], "datetimeDelivery")
            if payload.get("datetimeDelivery")
            else None
        )
    except ValueError as err:
        return jsonify({"message": str(err)}), 400

    request_model = Request(
        id_request=id_request,
        type_request=payload.get("typeRequest"),
        datetime_coming=parsed_coming,
        datetime_delivery=parsed_delivery,
        status=payload.get("status"),
        total_price=payload.get("totalPrice"),
    )

    items_payload: List[Dict[str, Any]] = payload.get("items") or []

    for item_data in items_payload:
        name = item_data.get("name")
        if not name:
            return jsonify({"message": 'Each request item must include "name"'}), 400

        part_number_raw = item_data.get("partNumber")
        if part_number_raw in (None, ""):
            part_number_value = None
        elif isinstance(part_number_raw, (int, float)):
            part_number_value = str(part_number_raw).strip()
        elif isinstance(part_number_raw, str):
            part_number_value = part_number_raw.strip() or None
        else:
            return jsonify({"message": 'Field "partNumber" must be a string or number if provided'}), 400

        serial_raw = item_data.get("serialNumber")
        serial_value = None
        if serial_raw not in (None, ""):
            if isinstance(serial_raw, (int, float)):
                serial_value = int(serial_raw)
            elif isinstance(serial_raw, str):
                serial_raw = serial_raw.strip()
                if serial_raw:
                    try:
                        serial_value = int(serial_raw)
                    except ValueError:
                        return jsonify({"message": 'Field "serialNumber" must be an integer if provided'}), 400
            else:
                return jsonify({"message": 'Field "serialNumber" must be an integer if provided'}), 400

        item_model = RequestItem(
            part_number=part_number_value,
            name=name,
            quantity=item_data.get("quantity"),
            unit=item_data.get("unit"),
            brand=item_data.get("brand"),
            model=item_data.get("model"),
            serial_number=serial_value,
            scheme=item_data.get("scheme"),
            pos_scheme=item_data.get("posScheme"),
            material=item_data.get("material"),
            comment=item_data.get("comment"),
            unit_price=item_data.get("unitPrice"),
            total_price=item_data.get("totalPrice"),
            product_id=item_data.get("productId"),
        )
        request_model.items.append(item_model)

    db.session.add(request_model)

    try:
        db.session.commit()
    except IntegrityError as err:
        db.session.rollback()
        return jsonify({"message": str(err)}), 400

    return jsonify(serialize_request(request_model)), 201
